trin-core/src/ethereum/bridge.rs

```diff
@@ -0,0 +1,227 @@
+use async_trait::async_trait;
+use ethers_core::types::{
+    Address, Block, Bytes, EIP1186ProofResponse, Transaction, TransactionReceipt, TxHash, H256,
+    U256, U64,
+};
+use ethers_providers::{
+    Ipc, JsonRpcClient, Middleware, Provider, ProviderError, PubsubClient, SubscriptionStream, Ws,
+};
+use log::error;
+use url::Url;
+
+use std::error::Error;
+use std::fmt;
+use std::path::Path;
+
+// TODO: Add more information to client and serialization errors.
+/// An error that can occur in communications with a bridge node.
+#[derive(Debug)]
+pub enum BridgeError {
+    EthClientError,
+    SerializationError,
+    ConnectionFailure,
+}
+
+impl fmt::Display for BridgeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::EthClientError => write!(f, "bridge ETH client error"),
+            Self::SerializationError => write!(f, "bridge serialization error"),
+            Self::ConnectionFailure => write!(f, "bridge connection failure"),
+        }
+    }
+}
+
+impl Error for BridgeError {}
+
+impl From<serde_json::Error> for BridgeError {
+    fn from(_: serde_json::Error) -> Self {
+        BridgeError::SerializationError
```

Comment:
I'm still new to rust so feel free to ignore this if it's misguided, but it looks like this (and the following method) throws away the original error. I bet that most times someone sees this error they'll be unsure what happened and they'll want to see the original error. I think implementing `source()` on `BridgeError` will make future problems easier to debug.

Reply:
Totally agree here. There is a `TODO` above the `BridgeError` definition to add more information to both the serialization and client error variants. We could either wrap the underlying errors or implement `source`. I'm not sure what is more idiomatic. I can go ahead and take this up and remove the `TODO` if we feel it necessary now.

Reply:
I'm not sure how high priority this is, so I would support implementing `source()` in this pr or create an issue to track it and handle in a subsequent pr.